Correction of Samsung Power OFF behaviour

RiRomain opened the pull request:
Addition of a delay after powering OFF a Samsung TV to avoid status
update to power the TV back ON.

## Description:
The Samsung TV component send a KEY to the TV to check the status of the TV (on/off).
This if made just after powering OFF the TV, turn some of the latest Samsung TV back ON (the TV stays connected to the network for a few seconds).

To fix this issue, I added a 15 seconds timer after the TV has been turned OFF during which no command are send to the TV.

**Related issue (if applicable):** fixes #10905

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#4123

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**

homeassistant commented:
Hi @RiRomain, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#10907).

Once you do that we will be able to review and accept this pull request.

Thanks!

MartinHjelmare commented on homeassistant/components/media_player/samsungtv.py:
```diff
@@ -171,13 +178,25 @@ def supported_features(self):
+        if self._powering_off_timer is not None\
+                and self._powering_off_timer.isAlive():
+            self._powering_off_timer.cancel()
+        self._powering_off_timer = Timer(15, _handle_powering_off_end)
```
Use `track_point_in_time` instead from helpers/event.py.

RiRomain replied:
```
    def _start_powering_off_timer(self):
        self._is_powering_off = True
        track_point_in_time(self.hass, self._handle_powering_off_end,
                            dt_util.utcnow() + timedelta(seconds=15))

    def _handle_powering_off_end(self):
        self._is_powering_off = False
```
With this code, the _handle_powering_off_end is never called. Do you see what I'm doing wrong?
I've checked, self.hass is not None.
Everything seems fine, but _handle_powering_off_end never run oO

MartinHjelmare replied:
The action function has to accept an argument, cause the time will be passed as an argument to the function when it is called. If you don't plan to use that argument, you can accept arbitrary arguments.
```py
def _handle_powering_off_end(self, *args):
    ...
```
Also, you have to separate utc time from local time, depending on which function you want to use. There are two different function for tracking time, `track_point_in_time` and `track_point_in_utc_time`. So if you use the former, you should capture the current time point in local time, `now()`.

RiRomain replied:
Thanks for your reply, it work after adding a second argument to my function.

In the end I implemented a simpler solution, I save the point in time at which occur the shutdown command:
`self._end_of_power_off_command  = dt_util.utcnow() + timedelta(seconds=15)`
And defined a function to check if we are still in a powering off phase:
```
    def _power_off_in_progress(self):
        return self._end_of_power_off_command is not None and \
               self._end_of_power_off_command > dt_util.utcnow()
```

I hope it's fine like that :smile:  (it's tested and working).

MartinHjelmare commented on homeassistant/components/media_player/samsungtv.py:
```diff
@@ -130,6 +133,10 @@ def get_remote(self):
     def send_key(self, key):
         """Send a key to the tv and handles exceptions."""
+        if self._is_powering_off and\
+                (key is 'KEY_POWER' or key is 'KEY_POWEROFF') is not True:
```
Compare strings with `==` and just prepend `not` to the expression in the parentheses, instead of `is not True`.

RiRomain replied:
Change applied.

MartinHjelmare commented on homeassistant/components/media_player/samsungtv.py:
```diff
@@ -130,6 +133,10 @@ def get_remote(self):
+        if self._is_powering_off and\
+                (key is 'KEY_POWER' or key is 'KEY_POWEROFF') is not True:
+            _LOGGER.info("TV is powering off, not sending command: %s", key)
+            return False
```
It doesn't look like anything is checking the return value of this method.

RiRomain replied:
```
    def update(self):
        """Retrieve the latest data."""
        # Send an empty key to see if we are still connected
        return self.send_key('KEY')
```
The update method return a value, but I think it doesn't need to. I will remove the update return statement, and also remove the return statement of send_key().
Is that ok?

MartinHjelmare replied:
Very good!

cgoudie commented:
+1

cgoudie commented:
This is better than what we have now, which is that we cant turn off the TV with home-assistant, so +1; howeverI dont think it will solve other methods of turning the tv off (remote, cec, etc...) those methods happen outside home-assistant, so I dont think they will detect self._is_powering_off state will they @RiRomain ?

RiRomain commented:
Thanks @MartinHjelmare, thanks for the review, I'll make the required change.
@cgoudie: yes you're right, that's still a problem. It however seems to happen less than when powering off with home-assistant. But a proper method to determine if the tv is off should be found (maybe polling the API endpoint, but I'm not sure that every Samsung dispose of the same endpoint)

Commit:
Correction of Samsung Power OFF behaviour

Addition of a delay after powering OFF a Samsung TV, this avoid status
update from powering the TV back ON.
Deletion of update() return statement, return value not used.

RiRomain commented:
@MartinHjelmare to avoid a mess in the history I made a force push, the new version is pushed and ready for a new review :smiley:

MartinHjelmare commented on homeassistant/components/media_player/samsungtv.py:
```diff
@@ -100,6 +102,9 @@ def __init__(self, host, port, name, timeout, mac):
         self._remote = None
+        # Mark the end of a shutdown command (need to wait 15 seconds before
+        # sending the next command to avoid turning the TV back ON).
+        self._end_of_power_off_command = None
```
Rename to `self._end_of_power_off`.

RiRomain replied:
Done.

MartinHjelmare commented:
We always squash when we merge, so it's better to not squash during the review process, to make it easier to follow the changes.

Commit:
Rename self._end_of_power_off_command into self._end_of_power_off

RiRomain commented:
Ok, didn't know about that. I'll avoid it in the future.

MartinHjelmare commented on homeassistant/components/media_player/samsungtv.py:
```diff
@@ -171,6 +183,9 @@ def supported_features(self):
 
     def turn_off(self):
         """Turn off media player."""
+        self._end_of_power_off \
```
I don't think you need to break this line anymore.

RiRomain replied:
You're right, I removed it.

Commit:
Removal of unused line break in Samsung TV component

MartinHjelmare reviewed (approved):
Thanks!
